pkg/client/interceptor/interceptor.go

```diff
@@ -2,5 +2,310 @@
+	// When specified, ContextName and Identity are ignored and retries are never attempted.
+	ServiceAccountBase64 string
+}
+
+// Interceptor is a GRPC interceptor that provides Unary and Stream client interceptors.
+type Interceptor struct {
+	initErr        error
+	userSigner     message.Signer
+	serviceAccount *serviceaccount.ServiceAccount
+	options        Options
+	initOnce       sync.Once
+	userSignerLock sync.Mutex
+	authEnabled    bool
+}
+
+// New creates a new client interceptor.
+func New(options Options) *Interceptor {
+	if options.InfoWriter == nil {
+		options.InfoWriter = os.Stderr
+	}
+
+	if options.AuthEnabledFunc == nil {
+		options.AuthEnabledFunc = func(ctx context.Context, cc *grpc.ClientConn) (bool, error) {
+			return true, nil
+		}
+	}
+
+	if options.GetUserKeyFunc == nil {
+		options.GetUserKeyFunc = func(ctx context.Context, cc *grpc.ClientConn, options *Options) (message.Signer, error) {
+			return options.UserKeyProvider.ReadValidKey(options.ContextName, options.Identity)
+		}
+	}
+
+	if options.RenewUserKeyFunc == nil {
+		options.RenewUserKeyFunc = renewUserKeyViaAuthFlow
+	}
+
+	return &Interceptor{
+		options: options,
+	}
+}
+
+// Unary returns a new unary client interceptor which signs requests.
+func (i *Interceptor) Unary() grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		return i.intercept(ctx, cc, method, func(ctx context.Context) error {
+			return invoker(ctx, method, req, reply, cc, opts...)
+		})
+	}
+}
+
+// Stream returns a new streaming client interceptor which signs requests.
+func (i *Interceptor) Stream() grpc.StreamClientInterceptor {
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		var stream grpc.ClientStream
+
+		err := i.intercept(ctx, cc, method, func(ctx context.Context) error {
+			var streamErr error
+
+			stream, streamErr = streamer(ctx, desc, cc, method, opts...)
+
+			return streamErr
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		return stream, nil
+	}
+}
+
+func (i *Interceptor) intercept(ctx context.Context, cc *grpc.ClientConn, method string, fn func(context.Context) error) error {
+	if ctx.Value(SkipInterceptorContextKey{}) != nil {
+		return fn(ctx)
+	}
+
+	ctx = context.WithValue(ctx, SkipInterceptorContextKey{}, struct{}{})
+
+	i.init(ctx, cc)
+
+	if i.initErr != nil {
+		return i.initErr
+	}
+
+	if !i.authEnabled {
+		return fn(ctx)
+	}
+
+	unsignedCtx := ctx
+	isRetryable := i.serviceAccount == nil
+
+	signedCtx, err := i.sign(unsignedCtx, cc, method)
+	if err != nil {
+		if isRetryable {
+			return i.retry(unsignedCtx, cc, method, err, fn)
+		}
+
+		return err
+	}
+
+	err = fn(signedCtx)
+	if status.Code(err) == codes.Unauthenticated && isRetryable {
+		return i.retry(unsignedCtx, cc, method, err, fn)
+	}
```

Comment:
This code is a bit hard to read with this `isRetryable` and `retry` calls in two places...
If we can unify the code we can also stop passing `firstAttemptErr` to the `retry` call.

Reply:
Maybe something like:
```golang
	var (
		signedCtx context.Context
		err error
	)

	callback := func() (bool, error) {
		signedCtx, err = i.sign(unsignedCtx, cc, method)
		if err != nil {
			return isRetryable, err
		}

		err = fn(signedCtx)
		if err != nil {
			return status.Code(err) == codes.Unauthenticated && isRetryable, err
		}
	}
	
Retry:
	retry, err = callback()

	if err != nil {
		if !retry {
			return err
		}

		fmt.Fprintf(i.options.InfoWriter, "Could not authenticate: %v\n", err)
		if err = i.renewUser(ctx, cc); err != nil {
			return err
		}

		isRetryable = false
		goto Retry
	}
	...
}

func (i *Interceptor) renewUser(ctx context.Context, cc *grpc.ClientConn) error {
	newSigner, err := i.options.RenewUserKeyFunc(ctx, cc, &i.options)
	if err != nil {
		return err
	}

	i.setUserSigner(newSigner)

	return nil
}
```

Reply:
Makes total sense, almost did this with the difference of using for loop instead of goto. Pls have a look.